### ISSUE TITLE
Why does hasActiveLibrary disable parallel builds for Haskell packages?

### ISSUE THREAD

**Issue by joelteon** [NixOS/nixpkgs#6709]
See https://github.com/NixOS/nixpkgs/blob/0fd3db4ba1d1d6d3a6108970ef96b24853593414/pkgs/development/haskell-modules/generic-builder.nix#L70

In practice, this just seems to build all Haskell libraries in serial except for the ones that only contain executables. I don't understand why this is necessary.

**Comment by peti** [comment]
This is necessary because parallel building greatly increases the chances triggering the non-determinism in `ghc`'s package id computations, i.e. with parallel building enabled it's near impossible to mix locally built libraries with ones downloaded from Hydra: a3ffe238e73f5cbf97a6982a88c4879659113145. @Fuuzetsu can probably tell you more about the technical details.

**Comment by Fuuzetsu** [comment]
Because of https://ghc.haskell.org/trac/ghc/ticket/4012 which with nix causes the package id problems if you fetch packages from different sources (such as build some packages locally and fetch some from Hydra). I think it should be safe to turn on parallel builds again for 7.10 though and see if there are any problems with that.

**Comment by wwat** [comment]
Ah, the irony; equal inputs should result in equal outputs for all the functional programming users, but for the authors those rules don't count. GHC is just not such a good compiler (because their type-checker is too complex, they are now afraid to touch it). Meanwhile, they are teaching others how to do software engineering. ROFL.

And of course the "if everything is pure, you can run it in parallel"-nonsense  (that fell flat on its face over a decade ago or so, invalidating pretty much all of their reason to exist). 

My solution for this ticket is to just point at laugh at the implementors at conferences and stick these points in their face and especially don't use their implementation until they fix it, but I have no doubt that you have other methods like begging them to fix it, cooperate, etc. When the Haskell people stop misrepresenting how great their little language is, I might stop laughing, but seriously... they have been implementing Haskell or similar languages for over two decades and they still didn't work out all the bugs?! Hilarious, if you ask me.

**Comment by Fuuzetsu** [comment]
@wwat Why even comment if you have nothing meaningful to say? No one asked you.

**Comment by copumpkin** [comment]
@wwat rants like that don't add anything to the discussion and will be removed. If you have a beef with how GHC is engineered and think it can be improved, take it up with its developers.

**Comment by joelteon** [comment]
@Fuuzetsu how do I test if parallel builds on 7.10 are ok?

**Comment by Fuuzetsu** [comment]
tl;dr: turn it on and watch for reports of breakage

There is no easy way: you can at best check that hashes generated by GHC for a package are being consistent (see comment 70 on the Trac) ticket. For nixpkgs what matters is that packages built on different machines generate the same package IDs so that we can distribute them. The IDs are always consistent if everything has been built on the same machine. The only way to test it on bigger scale is to turn it back on and see if it breaks eventually. Even pre-7.10 you could share build results between machines and often get lucky with the IDs and everything would work which makes it all the more difficult to detect.

**Comment by wwat** [comment]
What kind of a stupid losers are you that you feel the need to remove my comments? Go fuck yourself instead, thanks.

**Comment by copumpkin** [comment]
@wwat comments like that are prone to starting a flamewar in a place where it's entirely irrelevant. You don't just get to use our ticketing system as a place to whine and vent. If you want to whine and vent, do so elsewhere, or do something productive with your energy. You're welcome to post issues here, but your behavior so far on this thread (and a couple of others I've seen) is pretty inappropriate, and we'll not just leave it there to draw the flames. Start flamewars elsewhere if you must.

**Comment by peti** [comment]
I guess the question has been answered ...
